config/config_test.go

```diff
@@ -765,6 +765,79 @@ func TestMigrateConfigPurposeOneTreatment(t *testing.T) {
 	}
 }
 
+func TestMigrateConfigSpecialFeature1(t *testing.T) {
+	oldSpecialFeature1Config := []byte(`
+      gdpr:
+        tcf2:
+          special_purpose1:
+            enabled: true
+            vendor_exceptions: ["appnexus"]
+    `)
+	newSpecialFeature1Config := []byte(`
+      gdpr:
+        tcf2:
+          special_feature1:
+            enforce: true
+            vendor_exceptions: ["appnexus"]
+    `)
+	oldAndNewSpecialFeature1Config := []byte(`
+      gdpr:
+        tcf2:
+          special_purpose1:
+            enabled: false
+            vendor_exceptions: ["appnexus"]
+          special_feature1:
+            enforce: true
+            vendor_exceptions: ["rubicon"]
+    `)
+
+	tests := []struct {
+		description                         string
+		config                              []byte
+		wantSpecialFeature1Enforce          bool
+		wantSpecialFeature1VendorExceptions []string
+	}{
+		{
+			description: "New config and old config not set",
+			config:      []byte{},
+		},
+		{
+			description:                         "New config not set, old config set",
+			config:                              oldSpecialFeature1Config,
+			wantSpecialFeature1Enforce:          true,
+			wantSpecialFeature1VendorExceptions: []string{"appnexus"},
+		},
+		{
+			description:                         "New config set, old config not set",
+			config:                              newSpecialFeature1Config,
+			wantSpecialFeature1Enforce:          true,
+			wantSpecialFeature1VendorExceptions: []string{"appnexus"},
+		},
+		{
+			description:                         "New config and old config set",
+			config:                              oldAndNewSpecialFeature1Config,
+			wantSpecialFeature1Enforce:          true,
+			wantSpecialFeature1VendorExceptions: []string{"rubicon"},
+		},
+	}
+
+	for _, tt := range tests {
+		v := viper.New()
+		v.SetConfigType("yaml")
+		v.ReadConfig(bytes.NewBuffer(tt.config))
+
+		migrateConfigSpecialFeature1(v)
+
+		if len(tt.config) > 0 {
+			assert.Equal(t, tt.wantSpecialFeature1Enforce, v.Get("gdpr.tcf2.special_feature1.enforce").(bool), tt.description)
+			assert.Equal(t, tt.wantSpecialFeature1VendorExceptions, v.GetStringSlice("gdpr.tcf2.special_feature1.vendor_exceptions"), tt.description)
+		} else {
+			assert.Nil(t, v.Get("gdpr.tcf2.special_feature1.enforce"), tt.description)
+			assert.Nil(t, v.Get("gdpr.tcf2.special_feature1.vendor_exceptions"), tt.description)
+		}
```

Comment:
I find this to be a very comprehensive test case but: can we make one last assertion in the resulting `Configuration` object itself? I realize `TestSpecialFeature1VendorExceptionMap` does this to an extent but not when `oldSpecialFeature1Config` is set. Maybe we could add an `v.Unmarshal()` call at the end of this test case so we assert that `cfg.GDPR.TCF2.SpecialFeature1` was correctly set even when `gdpr.tcf2.special_purpose1` is set as in `oldSpecialFeature1Config`.
```
824     for _, tt := range tests {
825         v := viper.New()
826         v.SetConfigType("yaml")
827         v.ReadConfig(bytes.NewBuffer(tt.config))
828
829         migrateConfigSpecialFeature1(v)
830
831         if len(tt.config) > 0 {
832             assert.Equal(t, tt.wantSpecialFeature1Enforce, v.Get("gdpr.tcf2.special_feature1.enforce").(bool), tt.description)
833             assert.Equal(t, tt.wantSpecialFeature1VendorExceptions, v.GetStringSlice("gdpr.tcf2.special_feature1.vendor_exceptions"), tt.description)
834         } else {
835             assert.Nil(t, v.Get("gdpr.tcf2.special_feature1.enforce"), tt.description)
836             assert.Nil(t, v.Get("gdpr.tcf2.special_feature1.vendor_exceptions"), tt.description)
837         }
    +
    +	    var c Configuration
    +       err := v.Unmarshal(&c)
    +       assert.NoError(t, err, tt.desc)
    +
    +       // Assert expected values on the `cfg.GDPR.TCF2.SpecialFeature1` object itself
    +       assert.Equal(t, tt.wantSpecialFeature1Enforce, cfg.GDPR.TCF2.SpecialFeature1.Enforce, tt.desc)
    +       assert.ElementsMatch(t, tt.wantSpecialFeature1VendorExceptions, cfg.GDPR.TCF2.SpecialFeature1.VendorExceptions, tt.desc)
838     }
config/config_test.go
```
Or even better, a `New(v)` call so we can also assert the correctness of the `cfg.GDPR.TCF2.SpecialFeature1.VendorExceptionMap`
```
824     for _, tt := range tests {
825         v := viper.New()
826         v.SetConfigType("yaml")
827         v.ReadConfig(bytes.NewBuffer(tt.config))
828
829         migrateConfigSpecialFeature1(v)
830
831         if len(tt.config) > 0 {
832             assert.Equal(t, tt.wantSpecialFeature1Enforce, v.Get("gdpr.tcf2.special_feature1.enforce").(bool), tt.description)
833             assert.Equal(t, tt.wantSpecialFeature1VendorExceptions, v.GetStringSlice("gdpr.tcf2.special_feature1.vendor_exceptions"), tt.description)
834         } else {
835             assert.Nil(t, v.Get("gdpr.tcf2.special_feature1.enforce"), tt.description)
836             assert.Nil(t, v.Get("gdpr.tcf2.special_feature1.vendor_exceptions"), tt.description)
837         }
    +
    +	    cfg := New(v)
    +
    +       // Assert expected values on the `cfg.GDPR.TCF2.SpecialFeature1` object itself
    +       assert.Equal(t, tt.wantSpecialFeature1Enforce, cfg.GDPR.TCF2.SpecialFeature1.Enforce, tt.desc)
    +       assert.ElementsMatch(t, tt.wantSpecialFeature1VendorExceptions, cfg.GDPR.TCF2.SpecialFeature1.VendorExceptions, tt.desc)
    +       assert.Equal(t, tt.wantSpecialFeature1VendorExceptionsMap, cfg.GDPR.TCF2.SpecialFeature1.VendorExceptionMap, tt.desc)
838     }
config/config_test.go
```
Another option is to probably add an `oldSpecialFeature1Config` scenatio to `TestSpecialFeature1VendorExceptionMap`? What do you think?

Reply:
@guscarreon and I spoke offline about this.

My line of thinking was that the coverage was there across three unit tests but it was hard to see:
`TestMigrateConfigSpecialFeature1` —> verifies the deprecated special purpose config values are properly mapped to the new special feature config values
`TestFullConfig` —> verifies all of the special purpose struct fields, including the vendor exception map, are correctly populated based on what’s set in the special feature config values
`TestSpecialFeature1VendorExceptionMap` —> verifies the vendor exception map is created correctly no matter how many elements are in the vendor exception slice

I'm hesitant to make use of another function in the package to perform assertions in a pure unit test so instead of calling `New()` we agreed to call `unmarshal` as a happy medium to ensure what's written to the viper config in the migrate function is able to be unmarshaled into the struct.